### PR TITLE
Rewrite data source interface and internals

### PR DESCRIPTION
**Pull request by jacob-andersson**
Rewrite data source interface to look more like BarChart and LineChart. Adapted internals accordingly and refactored drawing logic. Further refactoring is going to be made to the drawing logic.

### CODE REVIEW

**Comment by 3lvis** [comment]
Massive refactoring is massive.

screen shot 2015-03-26 at 2 14 08 pm

**Comment by 3lvis** [comment]
@jacob-andersson do you mind updating the Example project?

**Comment by jacob-andersson** [comment]
@NSElvis Will do as soon as we have specified the last details and finished the implementation!

**Comment by jacob-andersson** [comment]
screen

File: Source/HYPScatterPlot.m [hunk 1]
```diff
@@ -28,136 +35,212 @@ - (instancetype)initWithFrame:(CGRect)frame
...
+    self.yAxisEndGradient = [UIColor clearColor];
+    self.defaultPointFillColor = [UIColor whiteColor];
+    self.selectedPointFillColor = [UIColor lightGrayColor];
+    self.selectedPointStrokeColor = [UIColor whiteColor];
+    self.axisLineWidth = HYPScatterPlotAxisLineWidth;
+    self.pointRadius = HYPScatterPlotCircleRadius;
+
+    UIView *leftYAxis = [UIView new];
+    leftYAxis.translatesAutoresizingMaskIntoConstraints = NO;
+    self.leftYAxisGradient = [CAGradientLayer layer];
```
Comment (attila-at-hyper) [inline Source/HYPScatterPlot.m]: Could you move initialization to a getter method? Also known as lazy initialization (which is a kind of standard here)? :)

File: Source/HYPScatterPlot.m [hunk 2]
```diff
@@ -28,136 +35,212 @@ - (instancetype)initWithFrame:(CGRect)frame
...
+    [leftYAxis.layer addSublayer:self.leftYAxisGradient];
+
+    UIView *rightYAxis = [UIView new];
+    rightYAxis.translatesAutoresizingMaskIntoConstraints = NO;
+    self.rightYAxisGradient = [CAGradientLayer layer];
```
Comment (attila-at-hyper) [inline Source/HYPScatterPlot.m]: See above.

File: Source/HYPScatterPlot.m [hunk 3]
```diff
@@ -28,136 +35,212 @@ - (instancetype)initWithFrame:(CGRect)frame
...
+    [rightYAxis.layer addSublayer:self.rightYAxisGradient];
+
+    self.minimumXLabel = [UILabel new];
```
Comment (attila-at-hyper) [inline Source/HYPScatterPlot.m]: ...and so on.

File: Source/HYPScatterPlot.m [hunk 4]
```diff
@@ -28,136 +35,212 @@ - (instancetype)initWithFrame:(CGRect)frame
...
+    self.maximumYLabel.attributedText = [self.dataSource maximumYValueFormattedInScatterPlotView:self];
 
-    CFRelease(frame);
+    if ([self.dataSource respondsToSelector:@selector(averageYValueFormattedInScatterPlotView:)]) {
+        self.averageYLabel.attributedText = [self.dataSource averageYValueFormattedInScatterPlotView:self];
+    }
 }
 
-- (void)drawAxes:(CGContextRef)context rect:(CGRect)rect
+- (void)updateYAxes
```
Comment (attila-at-hyper) [inline Source/HYPScatterPlot.m]: Is "axes" intentional? Maybe "axis" instead? (I was surprised to learn that "axes" is the plural form for "axis" though)

**Comment by attila-at-hyper** [comment]
I love this refactoring. @jacob-andersson, you rock. :+1:

**Comment by 3lvis** [comment]
@jacob-andersson Will that be included in this PR? Would be awful to merge and just break the Example.

**Comment by jacob-andersson** [comment]
@NSElvis No Pod exists yet for scatter, so I think it is okay!

Anyway, I think the interface is good enough now. Just want to do some more refactoring internally as well. Then I will update the example.

**Comment by 3lvis** [comment]
@jacob-andersson is it important that a Pod exists for us to not break existing code?

**Comment by attila-at-hyper** [comment]
Example needs to be updated once we're done with our demo.

**Comment by 3lvis** [comment]
@attila-at-hyper  
screen shot 2015-03-27 at 9 31 21 am

Don't forget to delete the branch :)

**Comment by 3lvis** [comment]
@jacob-andersson @attila-at-hyper made an issue so we don't forget 

https://github.com/hyperoslo/Scatter/issues/8 :)

**Comment by 3lvis** [comment]
I think we need to make sure to not break existing functionality when introducing code.
If this was important change and you didn't had time to fix the Demo you could reference the `branch` in your `Podfile`.

cc @hyperoslo/ios

**Pull request merged**